sips/cluster_consensus.md

```diff
@@ -28,143 +41,156 @@ The only validator-dependent field is `CommitteeIndex` and it does not have to b
-</p>
+}
 
-### New IDs
+type PartialSigRunner interface {
+	GetBeaconNode() BeaconNode
+	GetValCheckF() qbft.ProposedValueCheckF
+	GetSigner() types.BeaconSigner
+	GetNetwork() Network
 
-Since a single `QBFT instance` will be responsible for several validator duties, its ID must not be dependent on a validator key but rather on a cluster of operators. For that, we propose changing the `MessageID` from
+	//TODO should add local duty? or use the one in cd? 
+	UponDecided(cd *ConsesusData) error
+	ProcessPostConsensus(signedMsg *types.SignedPartialSignatureMessage) error
+}
 
-```mermaid
-flowchart RL
-	subgraph MessageID
-		Domain
-		ValidatorPublicKey
-		Role
-	end
+type PartialSigRunner struct
+	State          *State
+	Share          *types.Share
+	BeaconNetwork  types.BeaconNetwork
+	BeaconRoleType types.BeaconRole
 ```
 
-to
-
+#### Happy Flow
 
-```mermaid
-flowchart RL
-	subgraph MessageID
-		Domain
-		OperatorCluster
-		Role
-	end
-```
+1. `Cluster` receives all duties that match a certain slot, starts consensus for the relevant roles, and initializes the `PartialSigRunners` for the relevant Validators.
+2. `Cluster` receives consensus messages and hands them over to the `QBFTController` that has unchanged logic.
+3. Once `ProcessConsensus` returns `decided = true`, the `Cluster` will call `UponDecided(cd)` for each `PartialSigRunner` that has been initialized. This will cause an ommision of a partialSigMessage for each validator.
+4. `PartialSigRunner` will process post-consensus partial signature messages as before.
 
-### QBFT Controller
 
-The current `QBFTController` structure allows only one consensus instance at a time. This must change, extending the `QBFTController` into a router of messages for the different instances.
 
-### DutyRunner & QBFT Controller
+### ClusterID
 
-Since the `DutyRunner` will not have its specific `QBFTController`, it must have a way to start a consensus instance and receive its decided value. For that, we propose applying the observer design pattern by which the `DutyRunner` (observer) can be updated upon a `QBFT Instance` (observable) termination.
-
-We suggest that the `DutyRunner` holds a reference to the operator's `QBFTController` to start a `QBFT Instance` and observe it.
+An identifier for cluster must be added to `MessageID`.
 
 ```go
-func (r *DutyRunner) execute() {
-	r.QBFTController.StartConsensus(r.committee, r.duty, r)
-}
+[48]byte ClusterID
+
+// Return a 48 bytes ID for the cluster of operators
+func getClusterID(operatorIDs []OperatorID) ClusterID {
+	// Create a 16 bytes constant prefix for cluters
+	const prefix = []byte{0x00}
 
-func (c *QBFTController) StartConsensus(committee []types.Operator, duty types.Duty, observer Observer) {
-	if !c.HasConsensus(committee, duty) {
-		instance := c.StartInstance(committee, duty)
-	}
-	instance.registerObserver(observer)
+	// return the sha256 of the sortedIDs
+	return ClusterID(prefix + sha256.Sum256(bytes(sorted(operatorIDs))))
```

Comment:
could also be keccak
preferred a collision resistant function so no funny things will happen